velox/docs/develop/spilling.rst

```diff
@@ -504,4 +545,51 @@ spilling:
   partition number goes first.
 * To parallelize the hash table build from the spilled partition, the hash join
   bridge will split the spill partition files among the hash build operators
-  with each one having an equally-sized shard to restore.g
+  with each one having an equally-sized shard to restore.
+
+Future Work
+-----------
+Memory Arbitration
+^^^^^^^^^^^^^^^^^^
+Augment the memory management with memory arbitration support which kicks in to
+reclaim memory from running queries when any operator fails to allocate or
+reserve memory. The memory arbitrator can reclaim memory from both spillable
+and non-spillbale operators. For spillable operator, we need to add arena
+compaction to free up unused memory chunks. For non-spillable operator such as
+partial aggregation, the memory arbitrator can reclaim memory by triggering the
```

Comment:
> by triggering the partial aggregation state flush

It sound like the memory arbitration will have custom logic for flushing partial agg. I assume that's not the case. Rather PartialAgg operator will implement request to free up memory by flushing partial state.

Reply:
Yes, the operator just simply implement a reclaim API in their associated memory pool. The reclaim API will call into the operator's specific logic. For spillable operator, it triggers spilling internally but for non-spillable operator, it triggers partial aggregation state flush for partial aggregation or shuffle write for shuffle writer.